sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/FilterAnalyzer.scala

```diff
@@ -0,0 +1,186 @@
+    * @return cosmosFieldpath
+    */
+  private def canonicalCosmosFieldPath(sparkFilterColumnName: String): String = {
+    val result = new StringBuilder(FilterAnalyzer.rootName)
+    sparkFilterColumnName.split('.').foreach(cNamePart => result.append(s"['${normalizedFieldName(cNamePart)}']"))
+    result.toString
+  }
+
+  /**
+    * Parameter name in the parametrized query: e.g. @param1.
+    * @param paramNumber
+    * @return
+    */
+  private def paramName(paramNumber: Integer): String = {
+    s"@param$paramNumber"
+  }
+
+  // scalastyle:off cyclomatic.complexity
+  // scalastyle:off method.length
+  // scalastyle:off multiple.string.literals
+  private def appendCosmosQueryPredicate(queryBuilder: StringBuilder,
+                                         list: scala.collection.mutable.ListBuffer[(String, Any)],
+                                         filter: Filter): Boolean = {
+    val pName = paramName(list.size)
+    filter match {
+      case EqualTo(attr, value) =>
+        queryBuilder.append(canonicalCosmosFieldPath(attr)).append("=").append(pName)
+        list.append((pName, value))
+        true
+
+      case EqualNullSafe(attr, value) =>
+        // TODO moderakh check the difference between EqualTo and EqualNullSafe
+        queryBuilder.append(canonicalCosmosFieldPath(attr)).append("=").append(pName)
+        list.append((pName, value))
+        true
+
+      case LessThan(attr, value) =>
+        queryBuilder.append(canonicalCosmosFieldPath(attr)).append("<").append(pName)
+        list.append((pName, value))
+        true
+
+      case GreaterThan(attr, value) =>
+        queryBuilder.append(canonicalCosmosFieldPath(attr)).append(">").append(pName)
+        list.append((pName, value))
+        true
+
+      case LessThanOrEqual(attr, value) =>
+        queryBuilder.append(canonicalCosmosFieldPath(attr)).append("<=").append(pName)
+        list.append((pName, value))
+        true
+
+      case GreaterThanOrEqual(attr, value) =>
+        queryBuilder.append(canonicalCosmosFieldPath(attr)).append(">=").append(pName)
+        list.append((pName, value))
+        true
+
+      case In(attr, values) =>
+        queryBuilder.append(canonicalCosmosFieldPath(attr)).append(" IN ")
+        queryBuilder.append("(")
+        queryBuilder.append(
+          values.map(
+            value => {
+              val pName = paramName(list.size)
+              list.append((pName, value))
+              pName
+            }
+          ).mkString(","))
+        queryBuilder.append(")")
+        true
+
+      case StringStartsWith(attr, value: String) =>
+        queryBuilder.append("STARTSWITH(").append(canonicalCosmosFieldPath(attr)).append(pName).append(")")
+        list.append((pName, value))
+        true
+
+      case StringEndsWith(attr, value: String) =>
+        queryBuilder.append("ENDSWITH(").append(canonicalCosmosFieldPath(attr)).append(pName).append(")")
+        list.append((pName, value))
+        true
+
```

Comment:
Should we really push down EndsWith and Contains? I would not push down any filters which can not be served via full fidelity indexes - just costs more RU for the additional CPU usage when scanning through the documents

Reply:
will check with Brandon on this.

Reply:
I think we should push down all the filters (even if we can't serve them well from indexes). Even if spark can serve them faster and in an RU free manner doing it client side will probably cost more RUs in the long run. If we leave out a filter it causes the engine to have to return more documents, which drives up the RU cost the query. In most cases the cost of returning extra documents will outweigh the cost of scanning to prune the result set. 

In the case of string system functions like EndsWith and Contains the operations utilize the index fairly well. We run the system function on the index postings instead of doing a full scan on the dataset. This essentially makes it so that we are running the system function once per unique field, instead of once per document. We also make sure to run these filters after other filters that are quicker to run and will prune the dataset. Eventually when we have regex indexes these filters should be blazingly fast (and cheap).